edx_name_affirmation/signals.py

```diff
@@ -0,0 +1,179 @@
+# pylint: disable=logging-format-interpolation
+
+""""
+Signal handlers for IDV and proctoring service signals
+"""
+
+import logging
+
+from django.contrib.auth import get_user_model
+
+from edx_name_affirmation.models import VerifiedName
+from edx_name_affirmation.statuses import VerifiedNameStatus
+from edx_name_affirmation.toggles import is_verified_name_enabled
+
+try:
+    from lms.djangoapps.verify_student.signals import idv_update_signal
+except ImportError:
+    idv_update_signal = None
+
+try:
+    from edx_proctoring.utils import exam_attempt_status_signal
+except ImportError:
+    exam_attempt_status_signal = None
+
+User = get_user_model()
+
+log = logging.getLogger(__name__)
+
+
+def update_verified_name_for_idv_attempt(attempt_id, user_id, status, full_name, profile_name, **kwargs):
+    """
+    Receiver for IDV attempt updates
+
+    Args:
+        attempt_id(int): ID associated with the IDV attempt
+        user_id(int): ID associated with the IDV attempt's user
+        status(str): status in IDV language for the IDV attempt
+        full_name(str): name to be used as verified name
+        profile_name(str): user's current profile name
+    """
+    translated_status = VerifiedNameStatus.translate_idv_status(status)
+    log_str = ''
+
+    verified_name = VerifiedName.objects.filter(
+        user__id=user_id,
+        verification_attempt_id=attempt_id
+    ).order_by('-created').first()
+    if verified_name:
+        # if a verified name exists, update it if the status should trigger a transition
+        if translated_status:
+            verified_name.status = translated_status
+            verified_name.save()
+            log_str = 'Updated VerifiedName for user={user_id} with verification_attempt_id={attempt_id} to have ' \
+                      'status={status} '
+    else:
+        verified_name = VerifiedName.objects.filter(
+            user__id=user_id,
+            verified_name=full_name,
+            proctored_exam_attempt_id=None,
+            verification_attempt_id=None
+        ).order_by('-created').first()
+        if verified_name:
+            # if there is a verified name that matches the user and name,
+            # note the attempt id and update the status if needed
+            verified_name.verification_attempt_id = attempt_id
+            if translated_status:
+                verified_name.status = status
+            verified_name.save()
+            log_str = 'Updated VerifiedName for user={user_id} to have status={status} and ' \
+                      'verification_attempt_id={attempt_id} '
+        else:
+            # We shouldn't get to this case, but if there is no verified name associated with the attempt id or name,
+            # create a new verified name
```

Comment:
if we shouldn't get into this case, just patching it up seems like the wrong thing to do. Or if we do it we should log an error something like "we're doing this thing that the programmer thinks is impossible"

but isn't it part of the normal flow for IDV to make a new verified name and the comment is just not correct?

Reply:
I think the initial verified name record for an IDV attempt will not be created by a signal, and instead will be created by the work done in this ticket: https://openedx.atlassian.net/browse/MST-803. The reason for this is that the photo verification object for an IDV attempt would not be created until after the user has already submitted their IDV data, but we want to capture their verified name before entering the IDV flow. 

I'll add an error log for this case, because I think it's important that if we do reach this case that we don't drop the data on the floor.